External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -119,13 +188,107 @@ namespace FEXCore {
             }
         }
 
+        /**
+         * Generates a host-callable trampoline to call guest functions via the host ABI.
+         *
+         * This trampoline uses the same calling convention as the given HostPacker. Trampolines
+         * are cached, so it's safe to call this function repeatedly on the same arguments without
+         * leaking memory.
+         *
+         * Invoking the returned trampoline has the effect of:
+         * - packing the arguments (using the HostPacker identified by its SHA256)
+         * - performing a host->guest transition
+         * - unpacking the arguments via GuestUnpacker
+         * - calling the function at GuestTarget
+         *
+         * The primary use case of this is ensuring that guest function pointers ("callbacks")
+         * passed to thunked APIs can safely be called by the native host library.
+         */
+        static void MakeHostTrampolineForGuestFunction(void* ArgsRV) {
+          struct ArgsRV_t {
+              IR::SHA256Sum *HostPackerSha256;
+              uintptr_t GuestUnpacker;
+              uintptr_t GuestTarget;
+              uintptr_t rv; // Pointer to host trampoline + TrampolineInstanceInfo
+          } *args = reinterpret_cast<ArgsRV_t*>(ArgsRV);
+
+          LOGMAN_THROW_A_FMT(args->GuestTarget, "Tried to create host-trampoline to null pointer guest function");
+
+          const auto CTX = Thread->CTX;
+          const auto ThunkHandler = reinterpret_cast<ThunkHandler_impl *>(CTX->ThunkHandler.get());
+
+          const GuestcallInfo gci = { args->GuestUnpacker, args->GuestTarget };
+
+          // Try first with shared_lock
+          {
+            std::shared_lock lk(ThunkHandler->ThunksMutex);
+
+            auto found = ThunkHandler->GuestcallToHostTrampoline.find(gci);
+            if (found != ThunkHandler->GuestcallToHostTrampoline.end()) {
+              args->rv = found->second;
+              return;
+            }
+          }
+
+          std::lock_guard lk(ThunkHandler->ThunksMutex);
```

Comment:
This would be such a nice use case for a shared_lock with the ability to promote to unique_lock semantics.
Sadly we don't have one of these yet.